lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java

```diff
@@ -371,14 +426,26 @@ private void processFacetFields(
         }
 
         // Drill down:
-        int start;
-        if (ft.requireDimensionDrillDown) {
-          start = 1;
-        } else {
-          start = 2;
-        }
-        for (int i = start; i <= cp.length; i++) {
-          doc.add(new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+        if (ft.drillDownTermsIndexing != DrillDownTermsIndexing.NONE) {
+          StringField fullPathField =
+              new StringField(
+                  indexFieldName, pathToString(cp.components, cp.length), Field.Store.NO);
+          doc.add(fullPathField);
+
+          if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.DIMENSION_AND_FULL_PATH) {
+            doc.add(
+                new StringField(indexFieldName, pathToString(cp.components, 1), Field.Store.NO));
+          } else if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL_PATHS_NO_DIM) {
+            for (int i = 2; i < cp.length; i++) {
+              doc.add(
+                  new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+            }
+          } else if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL) {
+            for (int i = 1; i < cp.length; i++) {
+              doc.add(
+                  new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
+            }
+          }
```

Comment:
minor: you might consider the following, which might make it more clear to future readers that the only difference in these two options is the start length:

```suggestion
} else if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL_PATHS_NO_DIM || ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL) {
            int start;
            if (ft.drillDownTermsIndexing == DrillDownTermsIndexing.ALL) {
              start = 1;
            } else {
              start = 2;
            }
            for (int i = start; i < cp.length; i++) {
              doc.add(new StringField(indexFieldName, pathToString(cp.components, i), Field.Store.NO));
            }
          }
```

Reply:
hmmm tbh I actually preferred the existing one for readability as each option is checked once, whereas the suggested one has additional branching to check for one option twice. I see both styles having their pros & cons...

Reply:
Yeah, that's fair. My thinking was that the loop start being the only difference could be overlooked in a future change, but I don't really have strong opinions here. Thanks for considering a different approach :)